docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+| foo/bar\098506163554373999998PARIS  0.f9b82c166f695 |
+
+The following is a summary of the design and implementation details of bucket
+versioning, described first for the metadata engine and then for the API.
+
+## Implementation of Bucket Versioning in Metadata
+
+### Version Id Format
+
+The version id is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3. The following information is encoded before
+sending a response to a request. S3 also decodes the hexadecimal string
+received from a request before sending to metadata to retrieve a particular
+version.
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the texture format as metadata orders keys alphabetically
+
+The format of a `version_id` is:
+`ts` `rep_group_id` `seq_id`
+where:
+
+- `ts`: is the combination of epoch and an increasing number
+- `rep_group_id`: is the name of deployment(s) considered one unit used for
+   replication later
+- `seq_id`: is supposed to be a unique value based on metadata information.
+
+An example of a key: `foo\01234567890123456PARIS 1234.123456` indicating that
+this specific version of `foo` was the `123456`th entry created during the epoch
+`1234567890` in the replication group `PARIS` with `1234.1234566` as `seq_id`.
+
+### Master Version
+
+We store a copy of the latest version of an object's metadata using
+`object_name` as the key; this version is called the master version. The master
+versions of the objects facilitate the standard GET operation, which would
+otherwise need to scan among the list of versions of an object for its latest
+version.
+
+The following table shows the layout of all versions of `foo` in the first
+example stored in the metadata (with dot `.` representing the null separator):
+
+| key | value |
+|-----|:-----:|
+| foo |   B   |
+| foo.v2 |   B   |
+| foo.v1 |   A   |
+
+### Metadata Versioning Options
+
+S3 Server sends instructions to the metadata engine about whether to create a
+new version or overwrite, retrieve, or delete a specific version by sending
+values for special options in PUT, GET, or DELETE calls to metadata.
+
+Note: all operations (PUT and DELETE) that generate a new version of an object
+will return the `version_id` of the new version to the API.
+
+Note also that these describe only the basic CRUD operations that the metadata
+engine can handle. The process for creating new versions or delete markers has
+additional steps and is described more comprehensively in the section
+"Implementation of Bucket Versioning in API" and its subsections.
+
+#### PUT
+
+- no options: original PUT operation, will update the master version
+- `versioning: true` create a new version of the object,
```

Comment:
what happens if you have versioning: true and you send a versionID?  please specify.